// @caspertech/node-metaverse/dist/lib/enums/AssetTypeLL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetTypeLL = void 0;
var AssetTypeLL;
(function (AssetTypeLL) {
    AssetTypeLL[AssetTypeLL["texture"] = 0] = "texture";
    AssetTypeLL[AssetTypeLL["sound"] = 1] = "sound";
    AssetTypeLL[AssetTypeLL["callcard"] = 2] = "callcard";
    AssetTypeLL[AssetTypeLL["landmark"] = 3] = "landmark";
    AssetTypeLL[AssetTypeLL["script"] = 4] = "script";
    AssetTypeLL[AssetTypeLL["clothing"] = 5] = "clothing";
    AssetTypeLL[AssetTypeLL["object"] = 6] = "object";
    AssetTypeLL[AssetTypeLL["notecard"] = 7] = "notecard";
    AssetTypeLL[AssetTypeLL["category"] = 8] = "category";
    AssetTypeLL[AssetTypeLL["lsltext"] = 10] = "lsltext";
    AssetTypeLL[AssetTypeLL["lslbyte"] = 11] = "lslbyte";
    AssetTypeLL[AssetTypeLL["txtr_tga"] = 12] = "txtr_tga";
    AssetTypeLL[AssetTypeLL["bodypart"] = 13] = "bodypart";
    AssetTypeLL[AssetTypeLL["snd_wav"] = 17] = "snd_wav";
    AssetTypeLL[AssetTypeLL["img_tga"] = 18] = "img_tga";
    AssetTypeLL[AssetTypeLL["jpeg"] = 19] = "jpeg";
    AssetTypeLL[AssetTypeLL["animatn"] = 20] = "animatn";
    AssetTypeLL[AssetTypeLL["gesture"] = 21] = "gesture";
    AssetTypeLL[AssetTypeLL["simstate"] = 22] = "simstate";
    AssetTypeLL[AssetTypeLL["link"] = 24] = "link";
    AssetTypeLL[AssetTypeLL["link_f"] = 25] = "link_f";
    AssetTypeLL[AssetTypeLL["widget"] = 40] = "widget";
    AssetTypeLL[AssetTypeLL["person"] = 45] = "person";
    AssetTypeLL[AssetTypeLL["mesh"] = 49] = "mesh";
    AssetTypeLL[AssetTypeLL["settings"] = 56] = "settings";
})(AssetTypeLL = exports.AssetTypeLL || (exports.AssetTypeLL = {}));
