// @caspertech/node-metaverse/dist/lib/enums/AssetType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetType = void 0;
var AssetType;
(function (AssetType) {
    AssetType[AssetType["Unknown"] = -1] = "Unknown";
    AssetType[AssetType["Texture"] = 0] = "Texture";
    AssetType[AssetType["Sound"] = 1] = "Sound";
    AssetType[AssetType["CallingCard"] = 2] = "CallingCard";
    AssetType[AssetType["Landmark"] = 3] = "Landmark";
    AssetType[AssetType["Script"] = 4] = "Script";
    AssetType[AssetType["Clothing"] = 5] = "Clothing";
    AssetType[AssetType["Object"] = 6] = "Object";
    AssetType[AssetType["Notecard"] = 7] = "Notecard";
    AssetType[AssetType["Folder"] = 8] = "Folder";
    AssetType[AssetType["Category"] = 8] = "Category";
    AssetType[AssetType["LSLText"] = 10] = "LSLText";
    AssetType[AssetType["LSLBytecode"] = 11] = "LSLBytecode";
    AssetType[AssetType["TextureTGA"] = 12] = "TextureTGA";
    AssetType[AssetType["Bodypart"] = 13] = "Bodypart";
    AssetType[AssetType["SoundWAV"] = 17] = "SoundWAV";
    AssetType[AssetType["ImageTGA"] = 18] = "ImageTGA";
    AssetType[AssetType["ImageJPEG"] = 19] = "ImageJPEG";
    AssetType[AssetType["Animation"] = 20] = "Animation";
    AssetType[AssetType["Gesture"] = 21] = "Gesture";
    AssetType[AssetType["Simstate"] = 22] = "Simstate";
    AssetType[AssetType["Link"] = 24] = "Link";
    AssetType[AssetType["LinkFolder"] = 25] = "LinkFolder";
    AssetType[AssetType["MarketplaceFolder"] = 26] = "MarketplaceFolder";
    AssetType[AssetType["Widget"] = 40] = "Widget";
    AssetType[AssetType["Person"] = 46] = "Person";
    AssetType[AssetType["Mesh"] = 49] = "Mesh";
    AssetType[AssetType["Settings"] = 56] = "Settings";
})(AssetType = exports.AssetType || (exports.AssetType = {}));
