// @caspertech/node-metaverse/dist/lib/enums/Bumpiness.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bumpiness = void 0;
var Bumpiness;
(function (Bumpiness) {
    Bumpiness[Bumpiness["None"] = 0] = "None";
    Bumpiness[Bumpiness["Brightness"] = 1] = "Brightness";
    Bumpiness[Bumpiness["Darkness"] = 2] = "Darkness";
    Bumpiness[Bumpiness["Woodgrain"] = 3] = "Woodgrain";
    Bumpiness[Bumpiness["Bark"] = 4] = "Bark";
    Bumpiness[Bumpiness["Bricks"] = 5] = "Bricks";
    Bumpiness[Bumpiness["Checker"] = 6] = "Checker";
    Bumpiness[Bumpiness["Concrete"] = 7] = "Concrete";
    Bumpiness[Bumpiness["Crustytile"] = 8] = "Crustytile";
    Bumpiness[Bumpiness["Cutstone"] = 9] = "Cutstone";
    Bumpiness[Bumpiness["Discs"] = 10] = "Discs";
    Bumpiness[Bumpiness["Gravel"] = 11] = "Gravel";
    Bumpiness[Bumpiness["Petridish"] = 12] = "Petridish";
    Bumpiness[Bumpiness["Siding"] = 13] = "Siding";
    Bumpiness[Bumpiness["Stonetile"] = 14] = "Stonetile";
    Bumpiness[Bumpiness["Stucco"] = 15] = "Stucco";
    Bumpiness[Bumpiness["Suction"] = 16] = "Suction";
    Bumpiness[Bumpiness["Weave"] = 17] = "Weave";
})(Bumpiness = exports.Bumpiness || (exports.Bumpiness = {}));
