// @caspertech/node-metaverse/dist/lib/enums/DeRezDestination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeRezDestination = void 0;
var DeRezDestination;
(function (DeRezDestination) {
    DeRezDestination[DeRezDestination["AgentInventorySave"] = 0] = "AgentInventorySave";
    DeRezDestination[DeRezDestination["AgentInventoryCopy"] = 1] = "AgentInventoryCopy";
    DeRezDestination[DeRezDestination["TaskInventory"] = 2] = "TaskInventory";
    DeRezDestination[DeRezDestination["Attachment"] = 3] = "Attachment";
    DeRezDestination[DeRezDestination["AgentInventoryTake"] = 4] = "AgentInventoryTake";
    DeRezDestination[DeRezDestination["ForceToGodInventory"] = 5] = "ForceToGodInventory";
    DeRezDestination[DeRezDestination["TrashFolder"] = 6] = "TrashFolder";
    DeRezDestination[DeRezDestination["AttachmentToInventory"] = 7] = "AttachmentToInventory";
    DeRezDestination[DeRezDestination["AttachmentExists"] = 8] = "AttachmentExists";
    DeRezDestination[DeRezDestination["ReturnToOwner"] = 9] = "ReturnToOwner";
    DeRezDestination[DeRezDestination["ReturnToLastOwner"] = 10] = "ReturnToLastOwner";
})(DeRezDestination = exports.DeRezDestination || (exports.DeRezDestination = {}));
