// @caspertech/node-metaverse/dist/lib/enums/FolderType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FolderType = void 0;
var FolderType;
(function (FolderType) {
    FolderType[FolderType["None"] = -1] = "None";
    FolderType[FolderType["Texture"] = 0] = "Texture";
    FolderType[FolderType["Sound"] = 1] = "Sound";
    FolderType[FolderType["CallingCard"] = 2] = "CallingCard";
    FolderType[FolderType["Landmark"] = 3] = "Landmark";
    FolderType[FolderType["Clothing"] = 5] = "Clothing";
    FolderType[FolderType["Object"] = 6] = "Object";
    FolderType[FolderType["Notecard"] = 7] = "Notecard";
    FolderType[FolderType["Root"] = 8] = "Root";
    FolderType[FolderType["LSLText"] = 10] = "LSLText";
    FolderType[FolderType["BodyPart"] = 13] = "BodyPart";
    FolderType[FolderType["Trash"] = 14] = "Trash";
    FolderType[FolderType["Snapshot"] = 15] = "Snapshot";
    FolderType[FolderType["LostAndFound"] = 16] = "LostAndFound";
    FolderType[FolderType["Animation"] = 20] = "Animation";
    FolderType[FolderType["Gesture"] = 21] = "Gesture";
    FolderType[FolderType["Favorites"] = 23] = "Favorites";
    FolderType[FolderType["EnsembleStart"] = 26] = "EnsembleStart";
    FolderType[FolderType["EnsembleEnd"] = 45] = "EnsembleEnd";
    FolderType[FolderType["CurrentOutfit"] = 46] = "CurrentOutfit";
    FolderType[FolderType["Outfit"] = 47] = "Outfit";
    FolderType[FolderType["MyOutfits"] = 48] = "MyOutfits";
    FolderType[FolderType["Mesh"] = 49] = "Mesh";
    FolderType[FolderType["Inbox"] = 50] = "Inbox";
    FolderType[FolderType["Outbox"] = 51] = "Outbox";
    FolderType[FolderType["BasicRoot"] = 52] = "BasicRoot";
    FolderType[FolderType["MarketplaceListings"] = 53] = "MarketplaceListings";
    FolderType[FolderType["MarkplaceStock"] = 54] = "MarkplaceStock";
    FolderType[FolderType["Suitcase"] = 100] = "Suitcase";
})(FolderType = exports.FolderType || (exports.FolderType = {}));
