// @caspertech/node-metaverse/dist/lib/enums/InventoryType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InventoryType = void 0;
var InventoryType;
(function (InventoryType) {
    InventoryType[InventoryType["Unknown"] = -1] = "Unknown";
    InventoryType[InventoryType["Texture"] = 0] = "Texture";
    InventoryType[InventoryType["Sound"] = 1] = "Sound";
    InventoryType[InventoryType["CallingCard"] = 2] = "CallingCard";
    InventoryType[InventoryType["Landmark"] = 3] = "Landmark";
    InventoryType[InventoryType["Script"] = 4] = "Script";
    InventoryType[InventoryType["Clothing"] = 5] = "Clothing";
    InventoryType[InventoryType["Object"] = 6] = "Object";
    InventoryType[InventoryType["Notecard"] = 7] = "Notecard";
    InventoryType[InventoryType["Category"] = 8] = "Category";
    InventoryType[InventoryType["Folder"] = 8] = "Folder";
    InventoryType[InventoryType["RootCategory"] = 9] = "RootCategory";
    InventoryType[InventoryType["LSL"] = 10] = "LSL";
    InventoryType[InventoryType["Bodypart"] = 13] = "Bodypart";
    InventoryType[InventoryType["Snapshot"] = 15] = "Snapshot";
    InventoryType[InventoryType["Attachment"] = 17] = "Attachment";
    InventoryType[InventoryType["Wearable"] = 18] = "Wearable";
    InventoryType[InventoryType["Animation"] = 19] = "Animation";
    InventoryType[InventoryType["Gesture"] = 20] = "Gesture";
    InventoryType[InventoryType["Mesh"] = 22] = "Mesh";
    InventoryType[InventoryType["Widget"] = 23] = "Widget";
    InventoryType[InventoryType["Person"] = 24] = "Person";
    InventoryType[InventoryType["Settings"] = 25] = "Settings";
})(InventoryType = exports.InventoryType || (exports.InventoryType = {}));
