// @caspertech/node-metaverse/dist/lib/enums/MoneyTransactionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoneyTransactionType = void 0;
var MoneyTransactionType;
(function (MoneyTransactionType) {
    MoneyTransactionType[MoneyTransactionType["None"] = 0] = "None";
    MoneyTransactionType[MoneyTransactionType["FailSimulatorTimeout"] = 1] = "FailSimulatorTimeout";
    MoneyTransactionType[MoneyTransactionType["FailDataserverTimeout"] = 2] = "FailDataserverTimeout";
    MoneyTransactionType[MoneyTransactionType["ObjectClaim"] = 1000] = "ObjectClaim";
    MoneyTransactionType[MoneyTransactionType["LandClaim"] = 1001] = "LandClaim";
    MoneyTransactionType[MoneyTransactionType["GroupCreate"] = 1002] = "GroupCreate";
    MoneyTransactionType[MoneyTransactionType["ObjectPublicClaim"] = 1003] = "ObjectPublicClaim";
    MoneyTransactionType[MoneyTransactionType["GroupJoin"] = 1004] = "GroupJoin";
    MoneyTransactionType[MoneyTransactionType["TeleportCharge"] = 1100] = "TeleportCharge";
    MoneyTransactionType[MoneyTransactionType["UploadCharge"] = 1101] = "UploadCharge";
    MoneyTransactionType[MoneyTransactionType["LandAuction"] = 1102] = "LandAuction";
    MoneyTransactionType[MoneyTransactionType["ClassifiedCharge"] = 1103] = "ClassifiedCharge";
    MoneyTransactionType[MoneyTransactionType["ObjectTax"] = 2000] = "ObjectTax";
    MoneyTransactionType[MoneyTransactionType["LandTax"] = 2001] = "LandTax";
    MoneyTransactionType[MoneyTransactionType["LightTax"] = 2002] = "LightTax";
    MoneyTransactionType[MoneyTransactionType["ParcelDirFee"] = 2003] = "ParcelDirFee";
    MoneyTransactionType[MoneyTransactionType["GroupTax"] = 2004] = "GroupTax";
    MoneyTransactionType[MoneyTransactionType["ClassifiedRenew"] = 2005] = "ClassifiedRenew";
    MoneyTransactionType[MoneyTransactionType["GiveInventory"] = 3000] = "GiveInventory";
    MoneyTransactionType[MoneyTransactionType["ObjectSale"] = 5000] = "ObjectSale";
    MoneyTransactionType[MoneyTransactionType["Gift"] = 5001] = "Gift";
    MoneyTransactionType[MoneyTransactionType["LandSale"] = 5002] = "LandSale";
    MoneyTransactionType[MoneyTransactionType["ReferBonus"] = 5003] = "ReferBonus";
    MoneyTransactionType[MoneyTransactionType["InventorySale"] = 5004] = "InventorySale";
    MoneyTransactionType[MoneyTransactionType["RefundPurchase"] = 5005] = "RefundPurchase";
    MoneyTransactionType[MoneyTransactionType["LandPassSale"] = 5006] = "LandPassSale";
    MoneyTransactionType[MoneyTransactionType["DwellBonus"] = 5007] = "DwellBonus";
    MoneyTransactionType[MoneyTransactionType["PayObject"] = 5008] = "PayObject";
    MoneyTransactionType[MoneyTransactionType["ObjectPays"] = 5009] = "ObjectPays";
    MoneyTransactionType[MoneyTransactionType["GroupLandDeed"] = 6001] = "GroupLandDeed";
    MoneyTransactionType[MoneyTransactionType["GroupObjectDeed"] = 6002] = "GroupObjectDeed";
    MoneyTransactionType[MoneyTransactionType["GroupLiability"] = 6003] = "GroupLiability";
    MoneyTransactionType[MoneyTransactionType["GroupDividend"] = 6004] = "GroupDividend";
    MoneyTransactionType[MoneyTransactionType["GroupMembershipDues"] = 6005] = "GroupMembershipDues";
    MoneyTransactionType[MoneyTransactionType["ObjectRelease"] = 8000] = "ObjectRelease";
    MoneyTransactionType[MoneyTransactionType["LandRelease"] = 8001] = "LandRelease";
    MoneyTransactionType[MoneyTransactionType["ObjectDelete"] = 8002] = "ObjectDelete";
    MoneyTransactionType[MoneyTransactionType["ObjectPublicDecay"] = 8003] = "ObjectPublicDecay";
    MoneyTransactionType[MoneyTransactionType["ObjectPublicDelete"] = 8004] = "ObjectPublicDelete";
    MoneyTransactionType[MoneyTransactionType["LindenAdjustment"] = 9000] = "LindenAdjustment";
    MoneyTransactionType[MoneyTransactionType["LindenGrant"] = 9001] = "LindenGrant";
    MoneyTransactionType[MoneyTransactionType["LindenPenalty"] = 9002] = "LindenPenalty";
    MoneyTransactionType[MoneyTransactionType["EventFee"] = 9003] = "EventFee";
    MoneyTransactionType[MoneyTransactionType["EventPrize"] = 9004] = "EventPrize";
    MoneyTransactionType[MoneyTransactionType["StipendBasic"] = 10000] = "StipendBasic";
    MoneyTransactionType[MoneyTransactionType["StipendDeveloper"] = 10001] = "StipendDeveloper";
    MoneyTransactionType[MoneyTransactionType["StipendAlways"] = 10002] = "StipendAlways";
    MoneyTransactionType[MoneyTransactionType["StipendDaily"] = 10003] = "StipendDaily";
    MoneyTransactionType[MoneyTransactionType["StipendRating"] = 10004] = "StipendRating";
    MoneyTransactionType[MoneyTransactionType["StipendDelta"] = 10005] = "StipendDelta";
})(MoneyTransactionType = exports.MoneyTransactionType || (exports.MoneyTransactionType = {}));
