// @caspertech/node-metaverse/dist/lib/enums/ParcelFlags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParcelFlags = void 0;
var ParcelFlags;
(function (ParcelFlags) {
    ParcelFlags[ParcelFlags["None"] = 0] = "None";
    ParcelFlags[ParcelFlags["AllowFly"] = 1] = "AllowFly";
    ParcelFlags[ParcelFlags["AllowOtherScripts"] = 2] = "AllowOtherScripts";
    ParcelFlags[ParcelFlags["ForSale"] = 4] = "ForSale";
    ParcelFlags[ParcelFlags["AllowLandmark"] = 8] = "AllowLandmark";
    ParcelFlags[ParcelFlags["AllowTerraform"] = 16] = "AllowTerraform";
    ParcelFlags[ParcelFlags["AllowDamage"] = 32] = "AllowDamage";
    ParcelFlags[ParcelFlags["CreateObjects"] = 64] = "CreateObjects";
    ParcelFlags[ParcelFlags["ForSaleObjects"] = 128] = "ForSaleObjects";
    ParcelFlags[ParcelFlags["UseAccessGroup"] = 256] = "UseAccessGroup";
    ParcelFlags[ParcelFlags["UseAccessList"] = 512] = "UseAccessList";
    ParcelFlags[ParcelFlags["UseBanList"] = 1024] = "UseBanList";
    ParcelFlags[ParcelFlags["UsePassList"] = 2048] = "UsePassList";
    ParcelFlags[ParcelFlags["ShowDirectory"] = 4096] = "ShowDirectory";
    ParcelFlags[ParcelFlags["AllowDeedToGroup"] = 8192] = "AllowDeedToGroup";
    ParcelFlags[ParcelFlags["ContributeWithDeed"] = 16384] = "ContributeWithDeed";
    ParcelFlags[ParcelFlags["SoundLocal"] = 32768] = "SoundLocal";
    ParcelFlags[ParcelFlags["SellParcelObjects"] = 65536] = "SellParcelObjects";
    ParcelFlags[ParcelFlags["AllowPublish"] = 131072] = "AllowPublish";
    ParcelFlags[ParcelFlags["MaturePublish"] = 262144] = "MaturePublish";
    ParcelFlags[ParcelFlags["UrlWebPage"] = 524288] = "UrlWebPage";
    ParcelFlags[ParcelFlags["UrlRawHtml"] = 1048576] = "UrlRawHtml";
    ParcelFlags[ParcelFlags["RestrictPushObject"] = 2097152] = "RestrictPushObject";
    ParcelFlags[ParcelFlags["DenyAnonymous"] = 4194304] = "DenyAnonymous";
    ParcelFlags[ParcelFlags["LindenHome"] = 8388608] = "LindenHome";
    ParcelFlags[ParcelFlags["DenyTransacted"] = 16777216] = "DenyTransacted";
    ParcelFlags[ParcelFlags["AllowGroupScripts"] = 33554432] = "AllowGroupScripts";
    ParcelFlags[ParcelFlags["CreateGroupObjects"] = 67108864] = "CreateGroupObjects";
    ParcelFlags[ParcelFlags["AllowAPrimitiveEntry"] = 134217728] = "AllowAPrimitiveEntry";
    ParcelFlags[ParcelFlags["AllowGroupObjectEntry"] = 268435456] = "AllowGroupObjectEntry";
    ParcelFlags[ParcelFlags["AllowVoiceChat"] = 536870912] = "AllowVoiceChat";
    ParcelFlags[ParcelFlags["UseEstateVoiceChan"] = 1073741824] = "UseEstateVoiceChan";
    ParcelFlags[ParcelFlags["DenyAgeUnverified"] = 2147483648] = "DenyAgeUnverified";
})(ParcelFlags = exports.ParcelFlags || (exports.ParcelFlags = {}));
