// @caspertech/node-metaverse/dist/lib/enums/PrimFlags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrimFlags = void 0;
var PrimFlags;
(function (PrimFlags) {
    PrimFlags[PrimFlags["None"] = 0] = "None";
    PrimFlags[PrimFlags["Physics"] = 1] = "Physics";
    PrimFlags[PrimFlags["CreateSelected"] = 2] = "CreateSelected";
    PrimFlags[PrimFlags["ObjectModify"] = 4] = "ObjectModify";
    PrimFlags[PrimFlags["ObjectCopy"] = 8] = "ObjectCopy";
    PrimFlags[PrimFlags["ObjectAnyOwner"] = 16] = "ObjectAnyOwner";
    PrimFlags[PrimFlags["ObjectYouOwner"] = 32] = "ObjectYouOwner";
    PrimFlags[PrimFlags["Scripted"] = 64] = "Scripted";
    PrimFlags[PrimFlags["Touch"] = 128] = "Touch";
    PrimFlags[PrimFlags["ObjectMove"] = 256] = "ObjectMove";
    PrimFlags[PrimFlags["Money"] = 512] = "Money";
    PrimFlags[PrimFlags["Phantom"] = 1024] = "Phantom";
    PrimFlags[PrimFlags["InventoryEmpty"] = 2048] = "InventoryEmpty";
    PrimFlags[PrimFlags["JointHinge"] = 4096] = "JointHinge";
    PrimFlags[PrimFlags["JointP2P"] = 8192] = "JointP2P";
    PrimFlags[PrimFlags["JointLP2P"] = 16384] = "JointLP2P";
    PrimFlags[PrimFlags["JointWheel"] = 32768] = "JointWheel";
    PrimFlags[PrimFlags["AllowInventoryDrop"] = 65536] = "AllowInventoryDrop";
    PrimFlags[PrimFlags["ObjectTransfer"] = 131072] = "ObjectTransfer";
    PrimFlags[PrimFlags["ObjectGroupOwned"] = 262144] = "ObjectGroupOwned";
    PrimFlags[PrimFlags["ObjectYouOfficer"] = 524288] = "ObjectYouOfficer";
    PrimFlags[PrimFlags["CameraDecoupled"] = 1048576] = "CameraDecoupled";
    PrimFlags[PrimFlags["AnimSource"] = 2097152] = "AnimSource";
    PrimFlags[PrimFlags["CameraSource"] = 4194304] = "CameraSource";
    PrimFlags[PrimFlags["CastShadows"] = 8388608] = "CastShadows";
    PrimFlags[PrimFlags["DieAtEdge"] = 16777216] = "DieAtEdge";
    PrimFlags[PrimFlags["ReturnAtEdge"] = 33554432] = "ReturnAtEdge";
    PrimFlags[PrimFlags["Sandbox"] = 67108864] = "Sandbox";
    PrimFlags[PrimFlags["Flying"] = 134217728] = "Flying";
    PrimFlags[PrimFlags["ObjectOwnerModify"] = 268435456] = "ObjectOwnerModify";
    PrimFlags[PrimFlags["TemporaryOnRez"] = 536870912] = "TemporaryOnRez";
    PrimFlags[PrimFlags["Temporary"] = 1073741824] = "Temporary";
    PrimFlags[PrimFlags["ZlibCompressed"] = 2147483648] = "ZlibCompressed";
})(PrimFlags = exports.PrimFlags || (exports.PrimFlags = {}));
