// @caspertech/node-metaverse/dist/lib/enums/StatID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatID = void 0;
var StatID;
(function (StatID) {
    StatID[StatID["TimeDilation"] = 0] = "TimeDilation";
    StatID[StatID["FPS"] = 1] = "FPS";
    StatID[StatID["PhysFPS"] = 2] = "PhysFPS";
    StatID[StatID["AgentUPS"] = 3] = "AgentUPS";
    StatID[StatID["FrameMS"] = 4] = "FrameMS";
    StatID[StatID["NetMS"] = 5] = "NetMS";
    StatID[StatID["SimOtherMS"] = 6] = "SimOtherMS";
    StatID[StatID["SimPhysicsMS"] = 7] = "SimPhysicsMS";
    StatID[StatID["AgentMS"] = 8] = "AgentMS";
    StatID[StatID["ImagesMS"] = 9] = "ImagesMS";
    StatID[StatID["ScriptMS"] = 10] = "ScriptMS";
    StatID[StatID["NumTasks"] = 11] = "NumTasks";
    StatID[StatID["NumTasksActive"] = 12] = "NumTasksActive";
    StatID[StatID["NumAgentMain"] = 13] = "NumAgentMain";
    StatID[StatID["NumAgentChild"] = 14] = "NumAgentChild";
    StatID[StatID["NumScriptsActive"] = 15] = "NumScriptsActive";
    StatID[StatID["LSLIPS"] = 16] = "LSLIPS";
    StatID[StatID["InPPS"] = 17] = "InPPS";
    StatID[StatID["OutPPS"] = 18] = "OutPPS";
    StatID[StatID["PendingDownloads"] = 19] = "PendingDownloads";
    StatID[StatID["PendingUploads"] = 20] = "PendingUploads";
    StatID[StatID["VirtualSizeKB"] = 21] = "VirtualSizeKB";
    StatID[StatID["ResidentSizeKB"] = 22] = "ResidentSizeKB";
    StatID[StatID["PendingLocalUploads"] = 23] = "PendingLocalUploads";
    StatID[StatID["TotalUnackedBytes"] = 24] = "TotalUnackedBytes";
    StatID[StatID["PhysicsPinnedTasks"] = 25] = "PhysicsPinnedTasks";
    StatID[StatID["PhysicsLODTasks"] = 26] = "PhysicsLODTasks";
    StatID[StatID["SimPhysicsStepMS"] = 27] = "SimPhysicsStepMS";
    StatID[StatID["SimPhysicsShapeMS"] = 28] = "SimPhysicsShapeMS";
    StatID[StatID["SimPhysicsOtherMS"] = 29] = "SimPhysicsOtherMS";
    StatID[StatID["SimPhysicsMemory"] = 30] = "SimPhysicsMemory";
    StatID[StatID["ScriptEPS"] = 31] = "ScriptEPS";
    StatID[StatID["SimSpareTime"] = 32] = "SimSpareTime";
    StatID[StatID["SimSleepTime"] = 33] = "SimSleepTime";
    StatID[StatID["IOPumpTime"] = 34] = "IOPumpTime";
    StatID[StatID["PCTScriptsRun"] = 35] = "PCTScriptsRun";
    StatID[StatID["RegionIdle"] = 36] = "RegionIdle";
    StatID[StatID["RegionIdlePossible"] = 37] = "RegionIdlePossible";
    StatID[StatID["SimAIStepTimeMS"] = 38] = "SimAIStepTimeMS";
    StatID[StatID["SkippedAISilStepsPS"] = 39] = "SkippedAISilStepsPS";
    StatID[StatID["PCTSteppedCharacters"] = 40] = "PCTSteppedCharacters";
})(StatID = exports.StatID || (exports.StatID = {}));
