// @caspertech/node-metaverse/dist/lib/enums/WearableType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WearableType = void 0;
var WearableType;
(function (WearableType) {
    WearableType[WearableType["Shape"] = 0] = "Shape";
    WearableType[WearableType["Skin"] = 1] = "Skin";
    WearableType[WearableType["Hair"] = 2] = "Hair";
    WearableType[WearableType["Eyes"] = 3] = "Eyes";
    WearableType[WearableType["Shirt"] = 4] = "Shirt";
    WearableType[WearableType["Pants"] = 5] = "Pants";
    WearableType[WearableType["Shoes"] = 6] = "Shoes";
    WearableType[WearableType["Socks"] = 7] = "Socks";
    WearableType[WearableType["Jacket"] = 8] = "Jacket";
    WearableType[WearableType["Gloves"] = 9] = "Gloves";
    WearableType[WearableType["Undershirt"] = 10] = "Undershirt";
    WearableType[WearableType["Underpants"] = 11] = "Underpants";
    WearableType[WearableType["Skirt"] = 12] = "Skirt";
    WearableType[WearableType["Alpha"] = 13] = "Alpha";
    WearableType[WearableType["Tattoo"] = 14] = "Tattoo";
    WearableType[WearableType["Physics"] = 15] = "Physics";
    WearableType[WearableType["Invalid"] = 255] = "Invalid";
})(WearableType = exports.WearableType || (exports.WearableType = {}));
