// @caspertech/node-metaverse/dist/lib/classes/messages/AtomicPassObject.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtomicPassObjectMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class AtomicPassObjectMessage {
    constructor() {
        this.name = 'AtomicPassObject';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyHigh;
        this.id = Message_1.Message.AtomicPassObject;
    }
    getSize() {
        return 17;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.TaskData['TaskID'].writeToBuffer(buf, pos);
        pos += 16;
        buf.writeUInt8((this.TaskData['AttachmentNeedsSave']) ? 1 : 0, pos++);
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjTaskData = {
            TaskID: UUID_1.UUID.zero(),
            AttachmentNeedsSave: false
        };
        newObjTaskData['TaskID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjTaskData['AttachmentNeedsSave'] = (buf.readUInt8(pos++) === 1);
        this.TaskData = newObjTaskData;
        return pos - startPos;
    }
}
exports.AtomicPassObjectMessage = AtomicPassObjectMessage;
