// @caspertech/node-metaverse/dist/lib/classes/messages/CameraConstraint.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.CameraConstraintMessage = void 0;
const Vector4_1 = require("../Vector4");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class CameraConstraintMessage {
    constructor() {
        this.name = 'CameraConstraint';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.Zerocoded | MessageFlags_1.MessageFlags.FrequencyHigh;
        this.id = Message_1.Message.CameraConstraint;
    }
    getSize() {
        return 16;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.CameraCollidePlane['Plane'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjCameraCollidePlane = {
            Plane: Vector4_1.Vector4.getZero()
        };
        newObjCameraCollidePlane['Plane'] = new Vector4_1.Vector4(buf, pos);
        pos += 16;
        this.CameraCollidePlane = newObjCameraCollidePlane;
        return pos - startPos;
    }
}
exports.CameraConstraintMessage = CameraConstraintMessage;
