// @caspertech/node-metaverse/dist/lib/classes/messages/CompletePingCheck.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompletePingCheckMessage = void 0;
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class CompletePingCheckMessage {
    constructor() {
        this.name = 'CompletePingCheck';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyHigh;
        this.id = Message_1.Message.CompletePingCheck;
    }
    getSize() {
        return 1;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeUInt8(this.PingID['PingID'], pos++);
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjPingID = {
            PingID: 0
        };
        newObjPingID['PingID'] = buf.readUInt8(pos++);
        this.PingID = newObjPingID;
        return pos - startPos;
    }
}
exports.CompletePingCheckMessage = CompletePingCheckMessage;
