// @caspertech/node-metaverse/dist/lib/classes/messages/DataHomeLocationRequest.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataHomeLocationRequestMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class DataHomeLocationRequestMessage {
    constructor() {
        this.name = 'DataHomeLocationRequest';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.Zerocoded | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.DataHomeLocationRequest;
    }
    getSize() {
        return 24;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.Info['AgentID'].writeToBuffer(buf, pos);
        pos += 16;
        buf.writeUInt32LE(this.Info['KickedFromEstateID'], pos);
        pos += 4;
        buf.writeUInt32LE(this.AgentInfo['AgentEffectiveMaturity'], pos);
        pos += 4;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjInfo = {
            AgentID: UUID_1.UUID.zero(),
            KickedFromEstateID: 0
        };
        newObjInfo['AgentID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjInfo['KickedFromEstateID'] = buf.readUInt32LE(pos);
        pos += 4;
        this.Info = newObjInfo;
        const newObjAgentInfo = {
            AgentEffectiveMaturity: 0
        };
        newObjAgentInfo['AgentEffectiveMaturity'] = buf.readUInt32LE(pos);
        pos += 4;
        this.AgentInfo = newObjAgentInfo;
        return pos - startPos;
    }
}
exports.DataHomeLocationRequestMessage = DataHomeLocationRequestMessage;
