// @caspertech/node-metaverse/dist/lib/classes/messages/DeclineCallingCard.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeclineCallingCardMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class DeclineCallingCardMessage {
    constructor() {
        this.name = 'DeclineCallingCard';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.DeclineCallingCard;
    }
    getSize() {
        return 48;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.AgentData['AgentID'].writeToBuffer(buf, pos);
        pos += 16;
        this.AgentData['SessionID'].writeToBuffer(buf, pos);
        pos += 16;
        this.TransactionBlock['TransactionID'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjAgentData = {
            AgentID: UUID_1.UUID.zero(),
            SessionID: UUID_1.UUID.zero()
        };
        newObjAgentData['AgentID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjAgentData['SessionID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.AgentData = newObjAgentData;
        const newObjTransactionBlock = {
            TransactionID: UUID_1.UUID.zero()
        };
        newObjTransactionBlock['TransactionID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.TransactionBlock = newObjTransactionBlock;
        return pos - startPos;
    }
}
exports.DeclineCallingCardMessage = DeclineCallingCardMessage;
