// @caspertech/node-metaverse/dist/lib/classes/messages/DenyTrustedCircuit.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.DenyTrustedCircuitMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class DenyTrustedCircuitMessage {
    constructor() {
        this.name = 'DenyTrustedCircuit';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.DenyTrustedCircuit;
    }
    getSize() {
        return 16;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.DataBlock['EndPointID'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjDataBlock = {
            EndPointID: UUID_1.UUID.zero()
        };
        newObjDataBlock['EndPointID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.DataBlock = newObjDataBlock;
        return pos - startPos;
    }
}
exports.DenyTrustedCircuitMessage = DenyTrustedCircuitMessage;
