// @caspertech/node-metaverse/dist/lib/classes/messages/DeRezAck.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeRezAckMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class DeRezAckMessage {
    constructor() {
        this.name = 'DeRezAck';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.DeRezAck;
    }
    getSize() {
        return 17;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.TransactionData['TransactionID'].writeToBuffer(buf, pos);
        pos += 16;
        buf.writeUInt8((this.TransactionData['Success']) ? 1 : 0, pos++);
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjTransactionData = {
            TransactionID: UUID_1.UUID.zero(),
            Success: false
        };
        newObjTransactionData['TransactionID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjTransactionData['Success'] = (buf.readUInt8(pos++) === 1);
        this.TransactionData = newObjTransactionData;
        return pos - startPos;
    }
}
exports.DeRezAckMessage = DeRezAckMessage;
