// @caspertech/node-metaverse/dist/lib/classes/messages/EjectGroupMemberReply.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.EjectGroupMemberReplyMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class EjectGroupMemberReplyMessage {
    constructor() {
        this.name = 'EjectGroupMemberReply';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.EjectGroupMemberReply;
    }
    getSize() {
        return 33;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.AgentData['AgentID'].writeToBuffer(buf, pos);
        pos += 16;
        this.GroupData['GroupID'].writeToBuffer(buf, pos);
        pos += 16;
        buf.writeUInt8((this.EjectData['Success']) ? 1 : 0, pos++);
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjAgentData = {
            AgentID: UUID_1.UUID.zero()
        };
        newObjAgentData['AgentID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.AgentData = newObjAgentData;
        const newObjGroupData = {
            GroupID: UUID_1.UUID.zero()
        };
        newObjGroupData['GroupID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.GroupData = newObjGroupData;
        const newObjEjectData = {
            Success: false
        };
        newObjEjectData['Success'] = (buf.readUInt8(pos++) === 1);
        this.EjectData = newObjEjectData;
        return pos - startPos;
    }
}
exports.EjectGroupMemberReplyMessage = EjectGroupMemberReplyMessage;
