// @caspertech/node-metaverse/dist/lib/classes/messages/ForceObjectSelect.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForceObjectSelectMessage = void 0;
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class ForceObjectSelectMessage {
    constructor() {
        this.name = 'ForceObjectSelect';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.ForceObjectSelect;
    }
    getSize() {
        return ((4) * this.Data.length) + 2;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeUInt8((this.Header['ResetList']) ? 1 : 0, pos++);
        const count = this.Data.length;
        buf.writeUInt8(this.Data.length, pos++);
        for (let i = 0; i < count; i++) {
            buf.writeUInt32LE(this.Data[i]['LocalID'], pos);
            pos += 4;
        }
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjHeader = {
            ResetList: false
        };
        newObjHeader['ResetList'] = (buf.readUInt8(pos++) === 1);
        this.Header = newObjHeader;
        if (pos >= buf.length) {
            return pos - startPos;
        }
        const count = buf.readUInt8(pos++);
        this.Data = [];
        for (let i = 0; i < count; i++) {
            const newObjData = {
                LocalID: 0
            };
            newObjData['LocalID'] = buf.readUInt32LE(pos);
            pos += 4;
            this.Data.push(newObjData);
        }
        return pos - startPos;
    }
}
exports.ForceObjectSelectMessage = ForceObjectSelectMessage;
