// @caspertech/node-metaverse/dist/lib/classes/messages/InventoryAssetResponse.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.InventoryAssetResponseMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class InventoryAssetResponseMessage {
    constructor() {
        this.name = 'InventoryAssetResponse';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.InventoryAssetResponse;
    }
    getSize() {
        return 33;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.QueryData['QueryID'].writeToBuffer(buf, pos);
        pos += 16;
        this.QueryData['AssetID'].writeToBuffer(buf, pos);
        pos += 16;
        buf.writeUInt8((this.QueryData['IsReadable']) ? 1 : 0, pos++);
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjQueryData = {
            QueryID: UUID_1.UUID.zero(),
            AssetID: UUID_1.UUID.zero(),
            IsReadable: false
        };
        newObjQueryData['QueryID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjQueryData['AssetID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjQueryData['IsReadable'] = (buf.readUInt8(pos++) === 1);
        this.QueryData = newObjQueryData;
        return pos - startPos;
    }
}
exports.InventoryAssetResponseMessage = InventoryAssetResponseMessage;
