// @caspertech/node-metaverse/dist/lib/classes/messages/KillChildAgents.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.KillChildAgentsMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class KillChildAgentsMessage {
    constructor() {
        this.name = 'KillChildAgents';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.KillChildAgents;
    }
    getSize() {
        return 16;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.IDBlock['AgentID'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjIDBlock = {
            AgentID: UUID_1.UUID.zero()
        };
        newObjIDBlock['AgentID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.IDBlock = newObjIDBlock;
        return pos - startPos;
    }
}
exports.KillChildAgentsMessage = KillChildAgentsMessage;
