// @caspertech/node-metaverse/dist/lib/classes/messages/LiveHelpGroupRequest.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveHelpGroupRequestMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class LiveHelpGroupRequestMessage {
    constructor() {
        this.name = 'LiveHelpGroupRequest';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.LiveHelpGroupRequest;
    }
    getSize() {
        return 32;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.RequestData['RequestID'].writeToBuffer(buf, pos);
        pos += 16;
        this.RequestData['AgentID'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjRequestData = {
            RequestID: UUID_1.UUID.zero(),
            AgentID: UUID_1.UUID.zero()
        };
        newObjRequestData['RequestID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjRequestData['AgentID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.RequestData = newObjRequestData;
        return pos - startPos;
    }
}
exports.LiveHelpGroupRequestMessage = LiveHelpGroupRequestMessage;
