// @caspertech/node-metaverse/dist/lib/classes/messages/MergeParcel.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.MergeParcelMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class MergeParcelMessage {
    constructor() {
        this.name = 'MergeParcel';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.MergeParcel;
    }
    getSize() {
        return ((16) * this.SlaveParcelData.length) + 17;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.MasterParcelData['MasterID'].writeToBuffer(buf, pos);
        pos += 16;
        const count = this.SlaveParcelData.length;
        buf.writeUInt8(this.SlaveParcelData.length, pos++);
        for (let i = 0; i < count; i++) {
            this.SlaveParcelData[i]['SlaveID'].writeToBuffer(buf, pos);
            pos += 16;
        }
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjMasterParcelData = {
            MasterID: UUID_1.UUID.zero()
        };
        newObjMasterParcelData['MasterID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.MasterParcelData = newObjMasterParcelData;
        if (pos >= buf.length) {
            return pos - startPos;
        }
        const count = buf.readUInt8(pos++);
        this.SlaveParcelData = [];
        for (let i = 0; i < count; i++) {
            const newObjSlaveParcelData = {
                SlaveID: UUID_1.UUID.zero()
            };
            newObjSlaveParcelData['SlaveID'] = new UUID_1.UUID(buf, pos);
            pos += 16;
            this.SlaveParcelData.push(newObjSlaveParcelData);
        }
        return pos - startPos;
    }
}
exports.MergeParcelMessage = MergeParcelMessage;
