// @caspertech/node-metaverse/dist/lib/classes/messages/NearestLandingRegionUpdated.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.NearestLandingRegionUpdatedMessage = void 0;
const Long = require("long");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class NearestLandingRegionUpdatedMessage {
    constructor() {
        this.name = 'NearestLandingRegionUpdated';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.NearestLandingRegionUpdated;
    }
    getSize() {
        return 8;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeInt32LE(this.RegionData['RegionHandle'].low, pos);
        pos += 4;
        buf.writeInt32LE(this.RegionData['RegionHandle'].high, pos);
        pos += 4;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjRegionData = {
            RegionHandle: Long.ZERO
        };
        newObjRegionData['RegionHandle'] = new Long(buf.readInt32LE(pos), buf.readInt32LE(pos + 4));
        pos += 8;
        this.RegionData = newObjRegionData;
        return pos - startPos;
    }
}
exports.NearestLandingRegionUpdatedMessage = NearestLandingRegionUpdatedMessage;
