// @caspertech/node-metaverse/dist/lib/classes/messages/RegionHandleRequest.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegionHandleRequestMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class RegionHandleRequestMessage {
    constructor() {
        this.name = 'RegionHandleRequest';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.RegionHandleRequest;
    }
    getSize() {
        return 16;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.RequestBlock['RegionID'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjRequestBlock = {
            RegionID: UUID_1.UUID.zero()
        };
        newObjRequestBlock['RegionID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.RequestBlock = newObjRequestBlock;
        return pos - startPos;
    }
}
exports.RegionHandleRequestMessage = RegionHandleRequestMessage;
