// @caspertech/node-metaverse/dist/lib/classes/messages/ReportAutosaveCrash.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportAutosaveCrashMessage = void 0;
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class ReportAutosaveCrashMessage {
    constructor() {
        this.name = 'ReportAutosaveCrash';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.ReportAutosaveCrash;
    }
    getSize() {
        return 8;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeInt32LE(this.AutosaveData['PID'], pos);
        pos += 4;
        buf.writeInt32LE(this.AutosaveData['Status'], pos);
        pos += 4;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjAutosaveData = {
            PID: 0,
            Status: 0
        };
        newObjAutosaveData['PID'] = buf.readInt32LE(pos);
        pos += 4;
        newObjAutosaveData['Status'] = buf.readInt32LE(pos);
        pos += 4;
        this.AutosaveData = newObjAutosaveData;
        return pos - startPos;
    }
}
exports.ReportAutosaveCrashMessage = ReportAutosaveCrashMessage;
