// @caspertech/node-metaverse/dist/lib/classes/messages/SetCPURatio.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetCPURatioMessage = void 0;
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class SetCPURatioMessage {
    constructor() {
        this.name = 'SetCPURatio';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.SetCPURatio;
    }
    getSize() {
        return 1;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeUInt8(this.Data['Ratio'], pos++);
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjData = {
            Ratio: 0
        };
        newObjData['Ratio'] = buf.readUInt8(pos++);
        this.Data = newObjData;
        return pos - startPos;
    }
}
exports.SetCPURatioMessage = SetCPURatioMessage;
