// @caspertech/node-metaverse/dist/lib/classes/messages/SimulatorMapUpdate.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimulatorMapUpdateMessage = void 0;
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class SimulatorMapUpdateMessage {
    constructor() {
        this.name = 'SimulatorMapUpdate';
        this.messageFlags = MessageFlags_1.MessageFlags.Trusted | MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.SimulatorMapUpdate;
    }
    getSize() {
        return 4;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeUInt32LE(this.MapData['Flags'], pos);
        pos += 4;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjMapData = {
            Flags: 0
        };
        newObjMapData['Flags'] = buf.readUInt32LE(pos);
        pos += 4;
        this.MapData = newObjMapData;
        return pos - startPos;
    }
}
exports.SimulatorMapUpdateMessage = SimulatorMapUpdateMessage;
