// @caspertech/node-metaverse/dist/lib/classes/messages/StartPingCheck.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartPingCheckMessage = void 0;
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class StartPingCheckMessage {
    constructor() {
        this.name = 'StartPingCheck';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyHigh;
        this.id = Message_1.Message.StartPingCheck;
    }
    getSize() {
        return 5;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        buf.writeUInt8(this.PingID['PingID'], pos++);
        buf.writeUInt32LE(this.PingID['OldestUnacked'], pos);
        pos += 4;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjPingID = {
            PingID: 0,
            OldestUnacked: 0
        };
        newObjPingID['PingID'] = buf.readUInt8(pos++);
        newObjPingID['OldestUnacked'] = buf.readUInt32LE(pos);
        pos += 4;
        this.PingID = newObjPingID;
        return pos - startPos;
    }
}
exports.StartPingCheckMessage = StartPingCheckMessage;
