// @caspertech/node-metaverse/dist/lib/classes/messages/ViewerStartAuction.js
"use strict";
// This file has been automatically generated by writeMessageClasses.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewerStartAuctionMessage = void 0;
const UUID_1 = require("../UUID");
const MessageFlags_1 = require("../../enums/MessageFlags");
const Message_1 = require("../../enums/Message");
class ViewerStartAuctionMessage {
    constructor() {
        this.name = 'ViewerStartAuction';
        this.messageFlags = MessageFlags_1.MessageFlags.FrequencyLow;
        this.id = Message_1.Message.ViewerStartAuction;
    }
    getSize() {
        return 52;
    }
    // @ts-ignore
    writeToBuffer(buf, pos) {
        const startPos = pos;
        this.AgentData['AgentID'].writeToBuffer(buf, pos);
        pos += 16;
        this.AgentData['SessionID'].writeToBuffer(buf, pos);
        pos += 16;
        buf.writeInt32LE(this.ParcelData['LocalID'], pos);
        pos += 4;
        this.ParcelData['SnapshotID'].writeToBuffer(buf, pos);
        pos += 16;
        return pos - startPos;
    }
    // @ts-ignore
    readFromBuffer(buf, pos) {
        const startPos = pos;
        const newObjAgentData = {
            AgentID: UUID_1.UUID.zero(),
            SessionID: UUID_1.UUID.zero()
        };
        newObjAgentData['AgentID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        newObjAgentData['SessionID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.AgentData = newObjAgentData;
        const newObjParcelData = {
            LocalID: 0,
            SnapshotID: UUID_1.UUID.zero()
        };
        newObjParcelData['LocalID'] = buf.readInt32LE(pos);
        pos += 4;
        newObjParcelData['SnapshotID'] = new UUID_1.UUID(buf, pos);
        pos += 16;
        this.ParcelData = newObjParcelData;
        return pos - startPos;
    }
}
exports.ViewerStartAuctionMessage = ViewerStartAuctionMessage;
