// @caspertech/node-metaverse/dist/lib/classes/ClientEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientEvents = void 0;
const rxjs_1 = require("rxjs");
class ClientEvents {
    constructor() {
        this.onNearbyChat = new rxjs_1.Subject();
        this.onInstantMessage = new rxjs_1.Subject();
        this.onGroupInvite = new rxjs_1.Subject();
        this.onFriendRequest = new rxjs_1.Subject();
        this.onInventoryOffered = new rxjs_1.Subject();
        this.onLure = new rxjs_1.Subject();
        this.onTeleportEvent = new rxjs_1.Subject();
        this.onDisconnected = new rxjs_1.Subject();
        this.onCircuitLatency = new rxjs_1.Subject();
        this.onGroupChat = new rxjs_1.Subject();
        this.onGroupChatClosed = new rxjs_1.Subject();
        this.onGroupNotice = new rxjs_1.Subject();
        this.onGroupChatSessionJoin = new rxjs_1.Subject();
        this.onGroupChatAgentListUpdate = new rxjs_1.Subject();
        this.onFriendResponse = new rxjs_1.Subject();
        this.onInventoryResponse = new rxjs_1.Subject();
        this.onScriptDialog = new rxjs_1.Subject();
        this.onEventQueueStateChange = new rxjs_1.Subject();
        this.onFriendOnline = new rxjs_1.Subject();
        this.onFriendRights = new rxjs_1.Subject();
        this.onFriendRemoved = new rxjs_1.Subject();
        this.onPhysicsDataEvent = new rxjs_1.Subject();
        this.onParcelPropertiesEvent = new rxjs_1.Subject();
        this.onNewObjectEvent = new rxjs_1.Subject();
        this.onObjectUpdatedEvent = new rxjs_1.Subject();
        this.onObjectUpdatedTerseEvent = new rxjs_1.Subject();
        this.onObjectKilledEvent = new rxjs_1.Subject();
        this.onSelectedObjectEvent = new rxjs_1.Subject();
        this.onObjectResolvedEvent = new rxjs_1.Subject();
        this.onAvatarEnteredRegion = new rxjs_1.Subject();
        this.onRegionTimeDilation = new rxjs_1.Subject();
        this.onBulkUpdateInventoryEvent = new rxjs_1.Subject();
        this.onLandStatReplyEvent = new rxjs_1.Subject();
        this.onSimStats = new rxjs_1.Subject();
        this.onBalanceUpdated = new rxjs_1.Subject();
    }
}
exports.ClientEvents = ClientEvents;
