// @caspertech/node-metaverse/dist/lib/classes/public/LightData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightData = void 0;
const Color4_1 = require("../Color4");
class LightData {
    constructor(buf, pos, length) {
        this.Color = Color4_1.Color4.black;
        this.Radius = 0.0;
        this.Cutoff = 0.0;
        this.Falloff = 0.0;
        this.Intensity = 0.0;
        if (buf !== undefined && pos !== undefined && length !== undefined) {
            if (length >= 16) {
                this.Color = new Color4_1.Color4(buf, pos, false);
                pos += 4;
                this.Radius = buf.readFloatLE(pos);
                pos += 4;
                this.Cutoff = buf.readFloatLE(pos);
                pos += 4;
                this.Falloff = buf.readFloatLE(pos);
                if (typeof this.Color.alpha === 'number') {
                    this.Intensity = this.Color.alpha;
                }
                this.Color.alpha = 1.0;
            }
        }
    }
    writeToBuffer(buf, pos) {
        const tmpColour = new Color4_1.Color4(this.Color.getRed(), this.Color.getGreen(), this.Color.getBlue(), this.Color.getAlpha());
        tmpColour.alpha = this.Intensity;
        tmpColour.writeToBuffer(buf, pos);
        pos = pos + 4;
        buf.writeFloatLE(this.Radius, pos);
        pos = pos + 4;
        buf.writeFloatLE(this.Cutoff, pos);
        pos = pos + 4;
        buf.writeFloatLE(this.Falloff, pos);
    }
    getBuffer() {
        const buf = Buffer.allocUnsafe(16);
        this.writeToBuffer(buf, 0);
        return buf;
    }
}
exports.LightData = LightData;
